// @withkoji/vcc/dist/tools/SimpleEvent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _Handle = _interopRequireDefault(require("./Handle"));

var SimpleEvent = /*#__PURE__*/function () {
  function SimpleEvent() {
    (0, _classCallCheck2["default"])(this, SimpleEvent);
    this.handlers = undefined;
  }

  (0, _createClass2["default"])(SimpleEvent, [{
    key: "subscribe",
    value: function subscribe(handler) {
      var _this = this;

      var includeLast = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

      if (this.handlers == null) {
        this.handlers = [];
      }

      this.handlers.push(handler);

      if (this.lastEvent != null && includeLast) {
        handler(this.lastEvent);
      }

      return new _Handle["default"](function () {
        return _this.unsubscribe(handler);
      });
    }
  }, {
    key: "emit",
    value: function emit(event) {
      var _this2 = this;

      if (this.handlers != null) {
        this.handlers.forEach(function (handler) {
          handler(event, _this2.lastEvent);
        });
      }

      this.lastEvent = event;
    }
  }, {
    key: "unsubscribe",
    value: function unsubscribe(handler) {
      if (this.handlers == null) {
        return;
      }

      this.handlers = this.handlers.filter(function (h) {
        return h !== handler;
      });

      if (this.handlers.length === 0) {
        this.handlers = undefined;
      }
    }
  }]);
  return SimpleEvent;
}();

exports["default"] = SimpleEvent;
